### ISSUE TITLE
Mutate constant expressions

### ISSUE THREAD

**Issue by Danghor** [stryker-mutator/stryker-net#1018]
**Is your feature request related to a problem? Please describe.**
Currently, Stryker does not mutate constant expressions when assigned to fields

```csharp
const int foo = 5 - 1;
```

With #1017, it will also not support mutating enum member declarations:
```csharp
public enum Numbers
{
    One = 1,
    Two = (One + 1)
}
```

These syntax nodes are simply skipped, since Stryker will insert non-constant code during the mutation, which would cause code that does not compile:

```csharp
public enum Numbers
{
    One = 1,
    // CS0133 The expression being assigned to 'MutantOrchestratorTests.Numbers.Two' must be constant
    Two = StrykerXGJbRBlHxqRdD9O.MutantControl.IsActive(0) ? One + 1 : One - 1
}
```

**Describe the solution you'd like**
Stryker should mutate constant expressions in a way they compile, so that

```csharp
const int foo = 5 - 1;
```

becomes

```csharp
const int foo = 5 + 1;
```

This is a valid mutation that would break a piece of software and should be covered by unit tests. Therefore, it should be supported by Stryker to generate these kinds of mutations.

Probably deep changes are needed here, since generally, a ternary expression is inserted when mutating code. Instead, the code needs to be modified without inserting more function calls.

**Describe alternatives you've considered**
Instead of changing the mutation generation for all code, constant expressions could be handled in a different way, but that would lead to an inconsistent code base without much benefit.

**Additional context**
To avoid warnings until this feature is implemented, Issue #1016 and Pull Request #1017 have been created to skip EnumMemberDeclarationSyntax nodes. Otherwise, they would produce warnings when using Stryker on code containing those nodes.

**Comment by richardwerkman** [comment]
Thanks for your issue! Mutating constant values would indeed generate a lot of extra useful mutants. However, I'm not a big fan of your idea to change the way we place mutations. As you can read [here ](https://github.com/stryker-mutator/stryker-net/wiki/2.-Research)we have looked into other ways of placing mutants, and why we chose the ternary expressions. The main reason we want to keep it this way is: speed.

For the mutation:

``` cs
const int foo = 5 + 1;
```

Maybe we could just remove the const keyword. That way we can place the mutation like we always do. And the value would not change in the code since it was originally a const. So it will still behave like a const at runtime.

For your `enum` case that is not possible... However I doubt that we encounter a lot of possible mutations in enums. For me it's ok to skip enums. Unless someone finds an easy way to mutate this without having to mess a lot with the way we place mutants.

**Comment by Mobrockers** [comment]
What we could possibly do is introduce another variable that contains the mutated expression, then we could use ternary expressions wherever that variable is referenced to switch between the mutated variable and the original variable.

For enums we could introduce a completely new Enum in the same namespace. This enum can contain our 'mutated' constants. We can then use ternary expressions to switch between the original and mutated Enum wherever the Enum is referenced.

**Comment by richardwerkman** [comment]
How would that look in code? I'm not aware of a way to return another enum from a type conditionally. Or we would have to find all references to the enum (even in test projects) and switch between the original and mutated enum there.

**Comment by dupdob** [comment]
I think the easiest way to do this would be simply to build multiple assemblies.
I dont think switching variable names would work. Consider this
```csharp
 const int a = 0;
 const int b = a;
```
switching would mean something like
```csharp
const int a = 0;
const int a_mutated = 1;
const int b = MutantControl.IsActive(0) ? a_mutated : a; // won't build
```
Switching enum implies polymorphism.

But having one assembly per mutant take care of everything, at the expense of run time.
This could be considered as 'intense' mutation testing.

But IMHO, I am not sure this is a good direction. The point of Stryker is to generate the most interesting mutant, trying to find gaps in test. Not generating every possible alternate version of an assembly just so user can brag on how many mutants their tests kill.

**Comment by Mobrockers** [comment]
> switching would mean something like
> ```
> const int a = 0;
> const int a_mutated = 1;
> const int b = MutantControl.IsActive(0) ? a_mutated : a; // won't build
> ```
I did not think about constants being used to build other constants, that case would indeed not work.

> Or we would have to find all references to the enum (even in test projects) and switch between the original and mutated enum there.

Yep, I never said it would be easy, or even a good idea :p. Just that it is something that could potentially work.

**Comment by dupdob** [comment]
Just for the sake of the argument, on the opportunity for being silly:
> > switching would mean something like
> > 
> > ```const int a = 0;```
> > ```const int a_mutated = 1;```
> > ```const int b = MutantControl.IsActive(0) ? a_mutated : a; // won't build```
> > 
> 
> I did not think about constants being used to build other constants, that case would indeed not work. We would have to create new variables for every referencing const on the call stack and replace the references everywhere. Not pretty at all of course.

I am afraid I have to kill this idea right there. Let me give you another example. As usual, the simple code may look unrealistic, but my bet is that some real code could look like this:
```csharp
const int a = 1 + 1;
const int b = a + 1;
const int c = a + b;
```
The 'let's create alternate consts' leads to
```csharp
const int a = 1 + 1;
const int a_Mutant0 = 1 - 1;
const int b = a + 1;
const int b_Mutant1 = a - 1;
const int b_dependsOnMutant0 = a_Mutant0 + 1;
const int c = a + b ;
const int c_Mutant2 = a - b;
const int c_dependsOnMutant0 = a_Mutant0 + b_dependsOnMutant0;
const int c_dependsOnMutant1 = a + b_dependsOnMutant1;
```
I left the following example as an exercice for the reader.
```csharp
const int a = 1 + 1;
const int b = a + 1;
const int c = b + 1;
const int d = a + b + c;
```
 ⇒ I honestly think the only approach that may work in the real world is generating alternate assemblies.

**Comment by Mobrockers** [comment]
I agree the idea is horrible once you start working out the implications :p
